### ISSUE TITLE
Way to add in our own Context class

### ISSUE THREAD

**Issue by PythonCoderAS** [interactions-py/interactions.py#258]
I have a custom Context class that I use with `discord.ext.commands.Bot`, is there a way to do this as well in `discord_slash` for slash commands and/or components?

**Comment by LordOfPolls** [comment]
As of right now, no

May i ask why you need a custom context?

**Comment by PythonCoderAS** [comment]
I have a custom Context class to provide each command their own "logging" space that gets sent to a 3rd party when an error occurs.

**Comment by i0bs** [comment]
You should just be able to plug in a different `logging.Logger` [object here.](https://github.com/discord-py-slash-commands/discord-py-interactions/blob/master/discord_slash/context.py#L49) Direct instance modification would be `InteractionContext._logger` if you don't want to go through the hassle of having to re-call the entire class.

**Comment by PythonCoderAS** [comment]
I mentioned "logging", but that doesn't use the default Python logging. I am using a custom 3rd party solution that provides a web interface for errors as well as transactions to record the time taken for commands to execute.

**Comment by i0bs** [comment]
> I mentioned "logging", but that doesn't use the default Python logging.

For now, we only allow integration of our logger instances to use the logging module.

> I am using a custom 3rd party solution that provides a web interface for errors as well as transactions to record the time taken for commands to execute.

I think this would not be a necessary change to implement then for our library. There's no prevalent need as I would see it without needing to know if a lot of people would be opting to this kind of system, and it'd be optional as well if it were the case. For this kind of scenario it'd be best for you to modify the source as needed to adapt to your specific needs.

**Comment by LilSpazJoekp** [comment]
This is something that I would also like to see implement. I also use a custom class to house database, reddit, and logging interfaces (though this is through stdlib logging) for my bots.

**Comment by PythonCoderAS** [comment]
I'm not asking to implement said third-party actions into the code. My solution was to subclass the interactions client and manually add the attribute to every context instance created by the appropriate method, although that feels very hacky.

**Comment by i0bs** [comment]
This will be becoming a new feature in v4.0

**Comment by i0bs** [comment]
tl;dr: I'm dropping this idea.

There are a few reasons as to why, but the main overall one involves the complexity involved with how contextual classes will be structured and modeled based off of the Discord API. It's also counter-intuitive to be creating customized context classes that go against the nature of the API's given data and structure.

We also are making our transition from being an extension library of discord.py to being our own API wrapper/library detached entirely. It does not make sense for us to introduce something like this because of programming disconnection between us trying to support our users with the new codebase that is coming, and then the overall scheme.

This would also require an extensive overhaul with how we're already taking modeled information from the API and plugging it into the context class as well. A custom class would impose you would need to manually access, mutate and set all other given values. While that's technically the whole practicality of this feature request, the issue lies with incorrect data being given, which can break a lot of the models. Something as small as non-inclusively typed IDs of a Message, Guild or Channel for instance as a `str` and not `int` can easily cause a lot of problems. I also don't want to inclusively do a lot of things since I am keen to cater more towards strictly following the API's structure than to abstract it any further on my own end. It's a nice attribute to consider in library development, but I don't think it's overall worth the commitment to implement.

I'm more than happy for someone to add this as some other big PR, but I will not be coding this into v4.0 upon reconciliation.

**Comment by Xartrick** [comment]
> 
> 
> I mentioned "logging", but that doesn't use the default Python logging. I am using a custom 3rd party solution that provides a web interface for errors as well as transactions to record the time taken for commands to execute.

You can use different [handlers](https://docs.python.org/3/library/logging.handlers.html) to achieve what you want. This is still hacky but can still be done quite nicely if you create your own.

**Comment by PythonCoderAS** [comment]
Logging is one part of what I was trying to achieve, but streams won't fix all of my problems. I also wanted to instrument the commands ran, and it's going to be very hard without my own class, so that I can control attributes.

**Comment by i0bs** [comment]
We will be introducing an `interactions.ext` external framework for 3rd party library developers to introduce on the eventual release of v4.0 which is the best compromise I can see for both sides.
